### PR TITLE
ChoiceLayer, enable dict in explicit_search_sources

### PR DESCRIPTION
**Pull request by jvhoffbauer**
Enable usage of dicts in the `explicit_search_sources` of the choice layer. This is useful when writing custom logic in `custom_score_combine`, that behaves differently, depending on the inputs.

### CODE REVIEW

**Comment by jvhoffbauer** [comment]
Would it also be fine to accept None values in `explicit_search_sources`?

File: returnn/tf/layers/rec.py [hunk 1]
```diff
@@ -5010,7 +5017,11 @@ def get_dep_layers(self):
     # See also self.transform_config_dict where we might strip away the sources.
     ls = super(ChoiceLayer, self).get_dep_layers()
     if self.explicit_search_sources:
-      ls.extend(self.explicit_search_sources)
+      if isinstance(self.explicit_search_sources, dict):
+        additional_sources = self.explicit_search_sources.values()
```
Comment (albertz) [inline returnn/tf/layers/rec.py:5021]: This is non-deterministic in earlier Python versions.
Better:
```suggestion
        additional_sources = [src for _, src in sorted(self.explicit_search_sources.items())]
```

**Comment by albertz** [comment]
> Would it also be fine to accept None values in `explicit_search_sources`?

Yea, you can add that if you want.
Then you also need to extend `get_dep_layers` for that case.
```
additional_sources = [src for _, src in sorted(self.explicit_search_sources.items()) if src]
```

**Comment by patrick-wilken** [comment]
This is just to access the layers by key instead of index in `custom_score_combine`?

There is `assert isinstance(d["explicit_search_sources"], (list, tuple))` in `compile_tf_graph.py`, maybe we want to keep it up to date.

**Comment by jvhoffbauer** [comment]
Can we merge?

**Review by albertz — APPROVED** [review]
Looks good. Let's wait for the tests.

**Pull request merged**